### PR TITLE
 add initial version of VM and trace

### PR DESCRIPTION
**Pull request by hongyuanyang-uu**
1. add vm
2. add trace

### CODE REVIEW

**Review by spartucus — APPROVED** [review]

File: processor/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+const REGISTER_NUM: usize = 16;
+
+#[derive(Debug, Default)]
+pub struct Process {
+    pub clk: u32,
+    // todo: use felt
+    pub registers: [u32; REGISTER_NUM],
+    pub pc: usize,
+    // todo: fp stack need be modified for procedure
+    pub fp: Vec<usize>,
```
Comment (Softcloud88) [inline processor/src/lib.rs]: fp should be removed, maybe in later commits.

File: processor/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            value = self.registers[src_index];
+        }
+        value
+    }
+
+    pub fn execute(&mut self, program: &Program) -> Result<(), String> {
+        loop {
+            let ops_line =  program.instructions[self.pc].trim();
+            let ops: Vec<&str> = ops_line.split(' ').collect();
+            let mut opcode = ops.get(0).unwrap().to_lowercase();
```
Comment (Softcloud88) [inline processor/src/lib.rs:61]: compiled instruction structure might in a different form, should rewrite the parse logic.

File: processor/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = op1_value * op2_value;
+                    self.pc += 1;
+                }
+                "ret" => {
+                    debug!("opcode: ret");
+                    assert!(ops.len() == 1 ,  "ret params len is 0");
+                    self.pc = self.fp.pop().unwrap();
+                }
+                "end" => {
```
Comment (Softcloud88) [inline processor/src/lib.rs:181]: No end op.

File: processor/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                "mul" => {
+                    debug!("opcode: sub");
+                    assert!(ops.len() == 4 ,  "mul params len is 2");
+                    let dst_index = self.get_reg_index(ops.get(1).unwrap());
+                    let op1_value = self.get_index_value(ops.get(2).unwrap());
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = op1_value * op2_value;
+                    self.pc += 1;
+                }
+                "ret" => {
```
Comment (Softcloud88) [inline processor/src/lib.rs:172]: No fp, so ret may impl in another way?

**Review by Softcloud88 — APPROVED** [review]

File: processor/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                    self.pc = value.0;
+
+                }
+                "add" => {
+                    debug!("opcode: add");
+                    assert!(ops.len() == 4 ,  "add params len is 3");
+                    let dst_index = self.get_reg_index(ops.get(1).unwrap());
+                    let op1_index = self.get_reg_index(ops.get(2).unwrap());
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] =  self.registers[op1_index] + op2_value.0;
```
Comment (spartucus) [inline processor/src/lib.rs:142]: Need more code here: We can't do two array addition, and we also need consider overflow flag.

File: processor/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "sub" => {
+                    debug!("opcode: sub");
+                    assert!(ops.len() == 4 ,  "sub params len is 3");
+                    let dst_index = self.get_reg_index(ops.get(1).unwrap());
+                    let op1_index = self.get_reg_index(ops.get(2).unwrap());
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = self.registers[op1_index] - op2_value.0;
```
Comment (spartucus) [inline processor/src/lib.rs:154]: ditto!

File: processor/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                                              self.registers.clone(),self.flag, None);
+                    self.pc += 1;
+                }
+                "mul" => {
+                    debug!("opcode: sub");
+                    assert!(ops.len() == 4 ,  "mul params len is 3");
+                    let dst_index = self.get_reg_index(ops.get(1).unwrap());
+                    let op1_index = self.get_reg_index(ops.get(2).unwrap());
+                    let op2_value = self.get_index_value(ops.get(3).unwrap());
+                    self.registers[dst_index] = self.registers[op1_index] * op2_value.0;
```
Comment (spartucus) [inline processor/src/lib.rs:166]: ditto!

**Pull request closed without merging**